Fix bugs in _hydro_pwd

blt-r opened the pull request:
There were a couple of bugs in the implementation of `_hydro_pwd` which caused path to be rendered incorrectly:

- if there is a parent directory named the same way as git root, it will be "highlighted" instead of the actual git root.
For example, if you are in `/foo/repo/bar/repo/baz` and the second `repo` is git root, the first one will be highlighted.

- if you are in directory named `:` (but not in git repo) it will not be shown.
For example, if you are in `/tmp/:` you will see `/t/` instead of `/t/:`.

- if one of your parent directories is `:` and you are in git repo, the name of the repo will be shown instead of `:`.
For example, if you are in `/foo/:/bar/repo/baz` you will see `/f/repo/b/r/baz` instead of `/f/:/b/repo/baz`.

- if `fish_prompt_pwd_dir_length` is set to `0` and one of the parent directories starts with dot, the directory shown will start with dot.
For example, if you are in `/home/me/.config/fish` you will see `.fish` instead of `fish`.

- if you are in a directory that looks like your home directory but isn't, it will be replaced with `~`.
For example, if you are in `/tmp/home/me` you will see `/tmp~`.


(Sorry I didn't make a commit per bug fix, probably would have been more readable that way)

blt-r commented:
Also, it runs for every single prompt (or even twice per prompt?) instead of only when `$PWD` is changing. This is caused by `--on-variable fish_prompt_pwd_dir_length`. I know the standard `prompt_pwd` is setting the `$fish_prompt_pwd_dir_length`, and there is probably something else in fish that sets it. 

I would recommend not dealing with `$fish_prompt_pwd_dir_length` and instead have something like `$hydro_pwd_dir_length`

blt-r commented:
Just adder here another commit that I've also been using for past 8 months.

(see commit message for details)

For me, these two removed the small annoyances I had before.

jorgebucaran commented:
Since testing the code thoroughly would require significant effort and Hydro works well for me, I just haven't had the time to research this further. If you have the opportunity, we'd appreciate your feedback on this PR—even just to help add weight toward it being merged, hopefully soon.

blt-r commented:
So, having `--on-variable fish_prompt_pwd_dir_length` from `_hydro_pwd` causes the function to run unnecessarily. 

To see this, you can add "echo hydro pwd!" in the beginning of _hydro_pwd. Currently, for me, it results in:

```shell
hydro pwd!
~ ❯ hydro pwd!
~ ❯
hydro pwd!
~ ❯ hydro pwd!
~ ❯
hydro pwd!
~ ❯ hydro pwd!
~ ❯ cd Documents
hydro pwd!
hydro pwd!
hydro pwd!
~/Documents ❯ hydro pwd!
~/Documents ❯
hydro pwd!
~/Documents ❯ hydro pwd!
~/Documents ❯ ..
hydro pwd!
hydro pwd!
hydro pwd!
~ ❯ hydro pwd!
~ ❯
```

So it just runs twice per prompt for no reason and one more time when pwd is actually changed. If I then remove `--on-variable fish_prompt_pwd_dir_length`, I will see the following:

```shell
~ ❯
~ ❯
~ ❯
~ ❯ cd Documents
hydro pwd!
~/Documents ❯
~/Documents ❯
~/Documents ❯
~/Documents ❯ ..
hydro pwd!
~ ❯
~ ❯
~ ❯
```

Which is what is expected. So I removed `--on-variable fish_prompt_pwd_dir_length` from `_hydro_pwd`. 

Except finding the git root is done in `_hydro_pwd`, which is wrong, since user can init/deinit git repos, which will change the git root without changing the pwd. Really, finding the git root should be happening in `_hydro_postexec`, so I moved it there. You just never noticed that this is wrong because `_hydro_pwd` is running on every prompt anyway. 

Since `_hydro_pwd` doesn't depend on `fish_prompt_pwd_dir_length`, I changed hydro to use a new config variable `hydro_pwd_dir_length`, which is a breaking change. 

It is probably a good idea to make `hydro_pwd_dir_length` default to the value of `fish_prompt_pwd_dir_length` to maintain backwards compatibility, but I thought of that only just now.

While I was figuring all this out, I also cleaned up how the pwd string is actually computed, because the current implementation is hacky and more complicated than it needs to be. The main issue I encountered with it is this bug from original PR comment: 

> if `fish_prompt_pwd_dir_length` is set to `0` and one of the parent directories starts with dot, the directory shown will start with dot.
For example, if you are in `/home/me/.config/fish` you will see `.fish` instead of `fish`.

To simplify this process of constructing the pwd string, I factored it into a `_hydro_pretty_path` function, which just takes a normal path and does all the prettifying.

These two changes should ideally have been separate commits (or PRs), but they are weirdly separated into two commits, because I was doing them in parallel.

If you're interested, I can restructure commits in a way for them to make sense.
